File: scripts/extract_pdf3.js
const fs = require('fs');

const transcript_path = "C:\\Users\\Steven\\.gemini\\antigravity\\brain\\5f1aac63-eb77-44db-917d-a6dbc9626928\\.system_generated\\logs\\transcript.jsonl";
const out_path = "e:\\Gemini Projects\\PAASAU UCR UNA\\paasau\\temp_chunks\\raw_pdf3";

let pdf_content = "";
const lines_data = fs.readFileSync(transcript_path, 'utf8').split('\n');

for (const line of lines_data) {
  if (!line.trim()) continue;
  try {
    const data = JSON.parse(line);
    if (data.type === 'USER_INPUT' && data.content && data.content.includes('==Start of OCR')) {
      const regex = /==Start of OCR for page \d+==([\s\S]*?)==End of OCR for page \d+==/g;
      let match;
      while ((match = regex.exec(data.content)) !== null) {
        pdf_content += match[1] + "\n";
      }
    }
  } catch(e) {}
}

if (!pdf_content) {
  console.log("Could not find PDF OCR in transcript.");
  process.exit(1);
}

// Basic cleanup
const clean_lines = [];
for (let line of pdf_content.split('\n')) {
  line = line.trim();
  if (!line) continue;
  if (/^\d+\s+Instituto de Investigaciones Psicológicas$/.test(line)) continue;
  if (/^La Prueba de Aptitud Académica\s+\d+$/.test(line)) continue;
  if (["COMPRENSIÓN DE LECTURA", "COMPLETAR ORACIONES", "MATEMÁTICA", "CONOCIMIENTOS BÁSICOS", "― FIN DE LA PRÁC TI CA ―"].includes(line)) continue;
  clean_lines.push(line);
}

const text = "\n" + clean_lines.join("\n");
const questions = [];
let current_q_num = 1;

const pattern = /\n(\d+)\.\s/g;
let matches = [];
let match;
while ((match = pattern.exec(text)) !== null) {
  matches.push({ num: parseInt(match[1]), start: match.index, end: match.index + match[0].length });
}

for (let i = 0; i < matches.length; i++) {
  if (matches[i].num === current_q_num) {
    const start_idx = matches[i].end;
    const end_idx = (i + 1 < matches.length) ? matches[i+1].start : text.length;
    questions.push({ num: current_q_num, text: text.substring(start_idx, end_idx).trim() });
    current_q_num++;
  }
}

console.log(`Extracted ${questions.length} questions.`);

const parsed_questions = [];
for (const q of questions) {
  const opt_match = /\n\s*A\)\s+/.exec("\n" + q.text); // prepend newline to match first line A)
  if (opt_match) {
    const stem = q.text.substring(0, opt_match.index).trim();
    const opts_text = q.text.substring(opt_match.index);
    const opts = [];
    let current_opt = 'A';
    for (let j = 0; j < 5; j++) {
      const next_opt = String.fromCharCode(current_opt.charCodeAt(0) + 1);
      const start_rx = new RegExp('\\n\\s*' + current_opt + '\\)\\s+');
      const end_rx = new RegExp('\\n\\s*' + next_opt + '\\)\\s+');
      
      const start_m = start_rx.exec("\n" + opts_text);
      if (start_m) {
        const end_m = end_rx.exec("\n" + opts_text);
        if (end_m) {
          opts.push(opts_text.substring(start_m.index + start_m[0].length - 1, end_m.index - 1).trim());
        } else {
          opts.push(opts_text.substring(start_m.index + start_m[0].length - 1).trim());
        }
      } else {
        opts.push("");
      }
      current_opt = next_opt;
    }
    parsed_questions.push({ num: q.num, stem, opts });
  } else {
    parsed_questions.push({ num: q.num, stem: q.text, opts: [] });
  }
}

const answers = {
  1: 2, 2: 4, 3: 2, 4: 2, 5: 4, 6: 3, 7: 0, 8: 1, 9: 2, 10: 1,
  11: 3, 12: 4, 13: 1, 14: 0, 15: 0, 16: 3, 17: 4, 18: 4, 19: 2, 20: 3,
  21: 4, 22: 0, 23: 2, 24: 1, 25: 1, 26: 2, 27: 0, 28: 3, 29: 4, 30: 4,
  31: 0, 32: 3, 33: 2, 34: 1, 35: 4, 36: 2, 37: 2, 38: 4, 39: 1, 40: 3,
  41: 2, 42: 4, 43: 4, 44: 0, 45: 1, 46: 4, 47: 2, 48: 1, 49: 3, 50: 3,
  51: 1, 52: 4, 53: 2, 54: 3, 55: 2, 56: 3, 57: 2, 58: 2, 59: 3, 60: 4,
  61: 3, 62: 1, 63: 3, 64: 1, 65: 2, 66: 3, 67: 1, 68: 1, 69: 1, 70: 3,
  71: 0, 72: 2, 73: 2, 74: 1, 75: 1, 76: 2, 77: 1, 78: 4, 79: 3, 80: 1,
  81: 1, 82: 2, 83: 1, 84: 3, 85: 3
};

for (const pq of parsed_questions) {
  pq.ans = answers[pq.num] || 0;
  pq.domain = (pq.num <= 50) ? "verbal" : "math";
  pq.topic = "unknown";
}

const chunk_size = 22;
if (!fs.existsSync(out_path.substring(0, out_path.lastIndexOf('\\')))) {
  fs.mkdirSync(out_path.substring(0, out_path.lastIndexOf('\\')), { recursive: true });
}

for (let i = 0; i < Math.ceil(parsed_questions.length / chunk_size); i++) {
  const chunk = parsed_questions.slice(i * chunk_size, (i + 1) * chunk_size);
  fs.writeFileSync(`${out_path}_${i}.json`, JSON.stringify(chunk, null, 2), 'utf8');
  console.log(`Wrote chunk ${i} with ${chunk.length} questions.`);
}

File: scripts/extract_pdf3.py
import json
import re
import os

transcript_path = r"C:\Users\Steven\.gemini\antigravity\brain\5f1aac63-eb77-44db-917d-a6dbc9626928\.system_generated\logs\transcript_full.jsonl"
out_path = r"e:\Gemini Projects\PAASAU UCR UNA\paasau\temp_chunks\raw_pdf3.json"

pdf_content = ""
found_pdf = False

with open(transcript_path, 'r', encoding='utf-8') as f:
    for line in f:
        try:
            data = json.loads(line)
            if data.get('type') == 'USER_INPUT' and '==Start of PDF==' in data.get('content', ''):
                content = data['content']
                # Extract OCR blocks
                ocr_blocks = re.findall(r'==Start of OCR for page \d+==(.*?)==End of OCR for page \d+==', content, re.DOTALL)
                if ocr_blocks:
                    pdf_content = "\n".join(ocr_blocks)
                    found_pdf = True
        except Exception as e:
            pass

if not found_pdf:
    print("Could not find PDF OCR in transcript.")
    exit(1)

# Basic cleanup
lines = pdf_content.split('\n')
clean_lines = []
for line in lines:
    line = line.strip()
    if not line:
        continue
    # skip page numbers/headers like "20 Instituto de Investigaciones Psicológicas", "La Prueba de Aptitud Académica 21"
    if re.match(r'^\d+\s+Instituto de Investigaciones Psicológicas$', line):
        continue
    if re.match(r'^La Prueba de Aptitud Académica\s+\d+$', line):
        continue
    if line in ["COMPRENSIÓN DE LECTURA", "COMPLETAR ORACIONES", "MATEMÁTICA", "CONOCIMIENTOS BÁSICOS", "― FIN DE LA PRÁC TI CA ―"]:
        continue
    clean_lines.append(line)

text = "\n".join(clean_lines)

# Now, split by questions. Questions start with "1.", "2.", up to "85."
# Note: "I.", "II.", "III." might exist, so we only split on "\n\d+\.\s"
# We also have to be careful with decimals, but questions usually start on a newline.
# Let's find all indices of "\n\d+\." where the number is consecutive.
questions = []
current_q_text = ""
current_q_num = 1

# We add a newline to the start to match the first question
text = "\n" + text
pattern = re.compile(r'\n(\d+)\.\s')

matches = list(pattern.finditer(text))
for i in range(len(matches)):
    q_num_str = matches[i].group(1)
    q_num = int(q_num_str)
    
    # ensure it's actually the next question (or the first)
    if q_num == current_q_num:
        start_idx = matches[i].end()
        end_idx = matches[i+1].start() if i+1 < len(matches) else len(text)
        q_text = text[start_idx:end_idx].strip()
        questions.append({"num": q_num, "text": q_text})
        current_q_num += 1

print(f"Extracted {len(questions)} questions.")

# Now parse options A, B, C, D, E for each question
parsed_questions = []
for q in questions:
    q_text = q['text']
    # Split stem and options. Options start with " A) " or "A) "
    # We'll use regex to split
    opt_match = re.search(r'\n\s*A\)\s+', q_text)
    if opt_match:
        stem = q_text[:opt_match.start()].strip()
        opts_text = q_text[opt_match.start():]
        # Find B), C), D), E)
        opts = []
        for opt_letter in ['A', 'B', 'C', 'D', 'E']:
            next_letter = chr(ord(opt_letter) + 1)
            start_rx = re.compile(r'\n\s*' + opt_letter + r'\)\s+')
            end_rx = re.compile(r'\n\s*' + next_letter + r'\)\s+')
            
            start_m = start_rx.search(opts_text)
            if start_m:
                end_m = end_rx.search(opts_text)
                opt_start = start_m.end()
                if end_m:
                    opt_end = end_m.start()
                    opts.append(opts_text[opt_start:opt_end].strip())
                else:
                    opts.append(opts_text[opt_start:].strip())
            else:
                opts.append("") # Missing option?
        
        parsed_questions.append({
            "num": q["num"],
            "stem": stem,
            "opts": opts
        })
    else:
        # No options found?
        parsed_questions.append({
            "num": q["num"],
            "stem": q_text,
            "opts": []
        })

# Map answers based on manual extraction from the bubble sheet
# You verified these earlier
answers = {
    1: 2, 2: 4, 3: 2, 4: 2, 5: 4, 6: 3, 7: 0, 8: 1, 9: 2, 10: 1,
    11: 3, 12: 4, 13: 1, 14: 0, 15: 0, 16: 3, 17: 4, 18: 4, 19: 2, 20: 3,
    21: 4, 22: 0, 23: 2, 24: 1, 25: 1, 26: 2, 27: 0, 28: 3, 29: 4, 30: 4,
    31: 0, 32: 3, 33: 2, 34: 1, 35: 4, 36: 2, 37: 2, 38: 4, 39: 1, 40: 3,
    41: 2, 42: 4, 43: 4, 44: 0, 45: 1, 46: 4, 47: 2, 48: 1, 49: 3, 50: 3,
    51: 1, 52: 4, 53: 2, 54: 3, 55: 2, 56: 3, 57: 2, 58: 2, 59: 3, 60: 4,
    61: 3, 62: 1, 63: 3, 64: 1, 65: 2, 66: 3, 67: 1, 68: 1, 69: 1, 70: 3,
    71: 0, 72: 2, 73: 2, 74: 1, 75: 1, 76: 2, 77: 1, 78: 4, 79: 3, 80: 1,
    81: 1, 82: 2, 83: 1, 84: 3, 85: 3
}

for pq in parsed_questions:
    qnum = pq["num"]
    pq["ans"] = answers.get(qnum, 0)
    
    # Assign domain rough guess
    if 1 <= qnum <= 50:
        pq["domain"] = "verbal"
    else:
        pq["domain"] = "math"

# Write to chunks of 20
os.makedirs(os.path.dirname(out_path), exist_ok=True)
import math
chunk_size = 22 # To get exactly 4 chunks for 85 questions
for i in range(math.ceil(len(parsed_questions) / chunk_size)):
    chunk = parsed_questions[i*chunk_size : (i+1)*chunk_size]
    with open(out_path.replace(".json", f"_{i}.json"), 'w', encoding='utf-8') as f:
        json.dump(chunk, f, indent=2, ensure_ascii=False)
        print(f"Wrote chunk {i} with {len(chunk)} questions.")

